### PR TITLE
Add job and task code

### PR DESCRIPTION
**Pull request by chilianyi**

### CODE REVIEW

File: api/job.proto [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+	google.protobuf.StringValue directive = 8;
+	google.protobuf.StringValue executor = 9;
+	google.protobuf.UInt32Value task_count = 10;
+	google.protobuf.StringValue owner = 11;
+	google.protobuf.Timestamp create_time = 12;
+	google.protobuf.Timestamp status_time = 13;
+}
+message DescribeJobsRequest {
+	repeated string job_id = 1;
+	string cluster_id = 2;
```
Comment (zheng1) [inline api/job.proto]: cluster_id, app_id, app_version can be repeated

File: api/job.proto [hunk 2]
```diff
@@ -0,0 +1,94 @@
+// Copyright 2018 The OpenPitrix Authors. All rights reserved.
+// Use of this source code is governed by a Apache license
+// that can be found in the LICENSE file.
+
+syntax = "proto3";
+package openpitrix;
+
+// set go package name to pb
+option go_package = "pb";
```
Comment (chai2010) [inline api/job.proto]: openpitrix_pb is better than pb, we can use pb as the import alias name: `import pb "path"`

and we can use `package name` to set origin package name.
`option go_package` most use to set package import path (imported by other proto file).

File: api/job.proto [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+		};
+	};
+	schemes: HTTP;
+	consumes: "application/json";
+	produces: "application/json";
+};
+
+
+message CreateJobRequest {
+	string _ = 1; // placeholder for job_id
```
Comment (chai2010) [inline api/job.proto]: `_` is Go syntax, not proto file.

do u mean keep the `1` reserve?

File: api/job.proto [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+	string cluster_id = 2;
+	string app_id = 3;
+	string app_version = 4;
+	string job_action = 5;
+	string directive = 8;
```
Comment (chai2010) [inline api/job.proto]: why skip 6 and 7?

File: api/job.proto [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+message CreateJobResponse {
+	string job_id = 1;
+	string cluster_id = 2;
+	string app_id = 3;
+	string app_version = 4;
+}
+
+message Job {
+	google.protobuf.StringValue job_id = 1;
```
Comment (chai2010) [inline api/job.proto]: IMO, the empty job_id is invalid, why doe not use `string job_id = 1`?

PS: proto3's zero value is import feature, it is designed to fit Go's zero value.

File: api/job.proto [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+	google.protobuf.StringValue cluster_id = 2;
```
Comment (chai2010) [inline api/job.proto]: ditto

File: api/job.proto [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+	string app_id = 3;
+	string app_version = 4;
+	repeated string status = 5;
+	uint32 limit = 6;
+	uint32 offset = 7;
+}
+message DescribeJobsResponse {
+	uint32 total_count = 1;
+	repeated Job job_set = 2;
+}
```
Comment (chai2010) [inline api/job.proto]: the response is the slice, i think include `offset` will be more perfect

File: cmd/job-manager/main.go [hunk 8]
```diff
@@ -0,0 +1,16 @@
+// Copyright 2018 The OpenPitrix Authors. All rights reserved.
+// Use of this source code is governed by a Apache license
+// that can be found in the LICENSE file.
+
+// openpitrix runtime-env manager
```
Comment (chai2010) [inline cmd/job-manager/main.go]: job-manager?

File: cmd/task-manager/main.go [hunk 9]
```diff
@@ -0,0 +1,16 @@
+// Copyright 2018 The OpenPitrix Authors. All rights reserved.
+// Use of this source code is governed by a Apache license
+// that can be found in the LICENSE file.
+
+// openpitrix runtime-env manager
```
Comment (chai2010) [inline cmd/task-manager/main.go]: task-manager?

File: pkg/cmd/api/main.go [hunk 10]
```diff
@@ -35,9 +35,11 @@ func Serve() {
 
 	logger.Infof("Openpitrix %s\n", version.ShortVersion)
 	logger.Infof("App service http://%s:%s\n", constants.AppManagerHost, constants.AppManagerPort)
-	logger.Infof("Runtime service http://%s:%s\n", constants.RuntimeEnvManagerHost, constants.RuntimeEnvManagerPort)
+	logger.Infof("Runtime env service http://%s:%s\n", constants.RuntimeEnvManagerHost, constants.RuntimeEnvManagerPort)
```
Comment (chai2010) [inline pkg/cmd/api/main.go]: why `constants.XXXPort` is string type?

File: pkg/constants/common.go [hunk 11]
```diff
@@ -10,13 +10,17 @@ const RepoManagerHost = Prefix + "repo-manager"
...
+const TaskManagerHost = Prefix + "task-manager"
 const DbCtrlHost = Prefix + "db-ctrl"
 const ApiGatewayPort = "9100" // 91 is similar as Pi, Open[Pi]trix
 const RepoManagerPort = "9101"
 const AppManagerPort = "9102"
 const RuntimeEnvManagerPort = "9103"
 const ClusterManagerPort = "9104"
 const DbCtrlPort = "9105"
+const JobManagerPort = "9106"
+const TaskManagerPort = "9107"
```
Comment (chai2010) [inline pkg/constants/common.go]: why port is string type?

and use group block is better.

and the `Prefix` should not be public api, `"openpitrix-app-manager"` is better than `Prefix + "app-manager"`

File: pkg/manager/job/handler.go [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+	"openpitrix.io/openpitrix/pkg/pb"
+	"openpitrix.io/openpitrix/pkg/utils/sender"
+	"openpitrix.io/openpitrix/pkg/models"
+	"openpitrix.io/openpitrix/pkg/utils"
+	"openpitrix.io/openpitrix/pkg/db"
+)
+
+func (p *Server) CreateJob(ctx context.Context, req *pb.CreateJobRequest) (*pb.CreateJobResponse, error) {
+	s := sender.GetSenderFromContext(ctx)
+	logger.Debugf("Got req: %+v", req)
```
Comment (chai2010) [inline pkg/manager/job/handler.go]: move debug logger to the first line

File: pkg/manager/job/handler.go [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+	newJob := models.NewJob(req.GetClusterId(), req.GetAppId(), req.GetAppVersion(), req.GetJobAction(), req.GetDirective(), s.UserId)
```
Comment (chai2010) [inline pkg/manager/job/handler.go]: line is too long.

```go
	newJob := models.NewJob(
		req.GetClusterId(),
		req.GetAppId(),
		req.GetAppVersion(),
		req.GetJobAction(),
		req.GetDirective(),
		s.UserId,
	)
```

File: pkg/manager/job/handler.go [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+	var jobs []*models.Job
+	offset := utils.GetOffsetFromRequest(req)
+	limit := utils.GetLimitFromRequest(req)
+
+	query := p.db.
+		Select(models.JobColumns...).
+		From(models.JobTableName).
+		Offset(offset).
+		Limit(limit)
+	query = query.Where(db.Eq("owner", s.UserId))
```
Comment (chai2010) [inline pkg/manager/job/handler.go]: need blank line

File: api/job.proto (hunk 1 above)
Comment (chilianyi) [reply api/job.proto]: We may not query like that. 
We may query jobs belong to one cluster_id, or one app_id, or one app_version, but will not query jobs belong to multiple cluster_ids, I have not thought of the use of the scene.

File: api/job.proto (hunk 3 above)
Comment (chilianyi) [reply api/job.proto]: Yes, keep the 1 reserve, now job_id is useless because it is generate when handle the req.

File: api/job.proto (hunk 4 above)
Comment (chilianyi) [reply api/job.proto]: I will fix it.

File: api/job.proto (hunk 7 above)
Comment (chilianyi) [reply api/job.proto]: There is offset in req, no need in resp. 
If need slice, front end will pass offset and limit to do paging.

File: cmd/job-manager/main.go (hunk 8 above)
Comment (chilianyi) [reply cmd/job-manager/main.go]: I will fix it.

File: cmd/task-manager/main.go (hunk 9 above)
Comment (chilianyi) [reply cmd/task-manager/main.go]: I will fix it.

File: pkg/cmd/api/main.go (hunk 10 above)
Comment (chilianyi) [reply pkg/cmd/api/main.go]: Yes, int is better, i will change it.

File: pkg/constants/common.go (hunk 11 above)
Comment (chilianyi) [reply pkg/constants/common.go]: What do you mean by "group block". 
I will change `Prefix` to `prefix`. 
I think `Prefix + "app-manager"` is ok, through we will not change `openpitrix` prefix.

File: pkg/manager/job/handler.go (hunk 12 above)
Comment (chilianyi) [reply pkg/manager/job/handler.go]: OK

File: pkg/manager/job/handler.go (hunk 13 above)
Comment (chilianyi) [reply pkg/manager/job/handler.go]: OK

File: pkg/manager/job/handler.go (hunk 14 above)
Comment (chilianyi) [reply pkg/manager/job/handler.go]: OK

File: api/job.proto (hunk 2 above)
Comment (zheng1) [reply api/job.proto]: No, it's not necessary.

File: api/job.proto (hunk 1 above)
Comment (zheng1) [reply api/job.proto]: OK, `string` type in DescribeRequest should be `google.protobuf.StringValue`

File: api/job.proto (hunk 5 above)
Comment (zheng1) [reply api/job.proto]: In order to unify the Response, we will all the `string` in the Response writing `google.protobuf.StringValue`

File: pkg/constants/common.go (hunk 11 above)
Comment (zheng1) [reply pkg/constants/common.go]: I will refactor `constants` later, so this PR keep `constants` not changed

File: api/task.proto [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+message DescribeTasksRequest {
+	repeated string task_id = 1;
+	repeated string job_id = 2;
+	repeated string status = 3;
+	uint32 limit = 4; // uint64 marshal to json will convert to string
+	uint32 offset = 5;
+}
+message DescribeTasksResponse {
+	uint32 total_count = 1;
+	uint32 limit = 2;
```
Comment (zheng1) [inline api/task.proto]: `limit` is not necessary in response.

File: pkg/manager/job/handler.go [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+
+	// TODO: filter condition
```
Comment (zheng1) [inline pkg/manager/job/handler.go]: We can use `BuildFilterConditions` to refactor this now.

File: pkg/manager/job/handler.go [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+
+	count, err := query.Load(&jobs)
+	if err != nil {
+		// TODO: err_code should be implementation
+		return nil, status.Errorf(codes.Internal, "DescribeJobs: %+v", err)
+	}
+
+	res := &pb.DescribeJobsResponse{
+		JobSet:     models.JobsToPbs(jobs),
+		TotalCount: uint32(count),
```
Comment (zheng1) [inline pkg/manager/job/handler.go]: We can use `query.Count()` to count query now.

File: pkg/models/job.go [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+	Executor    string
+	TaskCount   uint32
+	CreateTime  time.Time
+	StatusTime  time.Time
+}
+
+var JobColumns = GetColumnsFromStruct(&Job{})
+
+func NewJob(clusterId, appId, appVersion, jobAction, directive, userId string) *Job {
+	return &Job{
```
Comment (zheng1) [inline pkg/models/job.go]: It looks like this line of code is not formatted.

File: api/task.proto (hunk 15 above)
Comment (chilianyi) [reply api/task.proto]: When user's limit is bigger than 200, then he/she should know the limit.

File: pkg/models/job.go (hunk 18 above)
Comment (chilianyi) [reply pkg/models/job.go]: Yes, it looks like `make build` not auto fmt code.

File: pkg/manager/job/handler.go (hunk 16 above)
Comment (chilianyi) [reply pkg/manager/job/handler.go]: Ok, I will pull new code and refactor.

File: api/task.proto (hunk 15 above)
Comment (zheng1) [reply api/task.proto]: There is no need to define in the Response, we can define it as a comment in Request. Such as:
```proto
message DescribeTasksRequest {
    // default is 20, max value is 200
    uint32 limit = 1;
    // default is 0
    uint32 offset = 2;
}
```
swagger-go will convert this comment to arguments' description.

Comment (chai2010) [reply api/task.proto]: if no limit in response, it should return error if `limit > 200` in req

Comment (chilianyi) [reply api/task.proto]: OK, I see we not return limit in qingcloud.

Comment (chilianyi) [reply api/task.proto]: @chai2010  No need to return error, just return ower largest limit record is ok.

**Comment by chai2010** [comment]
LGTM, wait for @zheng1

**Pull request merged**